<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        
        <link rel="canonical" href="https://lowandrew.github.io/SigSeekr/usage/">
        <link rel="shortcut icon" href="../img/favicon.ico">

	<title>Usage - SigSeekr</title>

        <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="../css/font-awesome-4.0.3.css" rel="stylesheet">
        <link rel="stylesheet" href="../css/highlight.css">
        <link href="../css/base.css" rel="stylesheet">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>

            <!-- Main title -->
            <a class="navbar-brand" href="..">SigSeekr</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            <!-- Main navigation -->
            <ul class="nav navbar-nav">
            
            
                <li >
                    <a href="..">Home</a>
                </li>
            
            
            
                <li >
                    <a href="../installation/">Installation</a>
                </li>
            
            
            
                <li class="active">
                    <a href="./">Usage</a>
                </li>
            
            
            </ul>

            <!-- Search, Navigation and Repo links -->
            <ul class="nav navbar-nav navbar-right">
                <li >
                    <a rel="next" href="../installation/">
                        <i class="fa fa-arrow-left"></i> Previous
                    </a>
                </li>
                <li class="disabled">
                    <a rel="prev" >
                        Next <i class="fa fa-arrow-right"></i>
                    </a>
                </li>
                
                <li>
                    <a href="https://github.com/lowandrew/SigSeekr">
                        
                            <i class="fa fa-github"></i>
                        
                        GitHub
                    </a>
                </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#quickstart">Quickstart</a></li>
        
            <li><a href="#detailed-usage">Detailed Usage</a></li>
        
    
    </ul>
</div></div>
            <div class="col-md-9" role="main">

<h1 id="quickstart">Quickstart</h1>
<p>If you want to run SigSeekr right away upon <a href="../installation/">installing</a> it, you can do so with a toy dataset.</p>
<p>This dataset is hosted on figshare - to get it, run the following command:</p>
<ul>
<li><code>wget https://ndownloader.figshare.com/files/9885379 &amp;&amp; tar xf 9885379</code></li>
</ul>
<p>You should now have a folder called <code>example-data</code> in your present working directory. To run SigSeekr, enter the following command:</p>
<ul>
<li><code>sigseekr.py -i example-data/inclusion/ -e example-data/exclusion/ -o sigseekr_output</code> </li>
</ul>
<p>The directory specified with the <code>-o</code> flag can be anything - it's the name of a directory where the output files will be created.
Upon entering the command, you should see output that is something like this:</p>
<pre><code class="bash"> [Elapsed Time: 0.00 seconds] Creating inclusion kmer set... 
 [Elapsed Time: 0.41 seconds] Creating exclusion kmer set... 
 [Elapsed Time: 0.83 seconds] Subtracting exclusion kmers from inclusion kmers with cutoff 1... 
 [Elapsed Time: 0.97 seconds] Found kmers unique to inclusion... 
 [Elapsed Time: 0.97 seconds] Generating contiguous sequences from inclusion kmers... 
 [Elapsed Time: 4.10 seconds] Removing unnecessary output files... 
 [Elapsed Time: 4.10 seconds] SigSeekr run complete! 
</code></pre>

<p>The <code>sigseekr_output</code> folder should have two files in it: <code>inclusion_kmers.fasta</code>, which lists all the kmers that are unique to the inclusion set, and <code>sigseekr_result.fasta</code>, which contains the regions that unique kmers span. In this case, <code>sigseekr_result.fasta</code> should have one unique region. To take a look at it, use the <code>cat</code> command:</p>
<ul>
<li><code>cat sigseekr_output/sigseekr_result.fasta</code></li>
</ul>
<p>The result that should come out of this is:</p>
<pre><code class="bash">&gt;contig1_sequence1
AACAGGCGACAGGCAGCATCACTAGCTACTA
</code></pre>

<h3 id="detailed-usage">Detailed Usage</h3>
<p>Detailed usage options can be found by typing <code>sigseekr.py --help</code>, which will give the following output. 
Further details on each option can be found below.</p>
<pre><code class="python">usage: sigseekr.py [-h] -i INCLUSION -e EXCLUSION -o OUTPUT_FOLDER
                   [-s KMER_SIZE] [-t THREADS] [-pcr] [-k]
                   [-p PLASMID_FILTERING] [-l]
                   [-a AMPLICON_SIZE [AMPLICON_SIZE ...]]

optional arguments:
  -h, --help            show this help message and exit
  -i INCLUSION, --inclusion INCLUSION
                        Path to folder containing genome(s) you want signature
                        sequences for. Genomes can be in FASTA or FASTQ
                        format. FASTA-formatted files should be uncompressed,
                        FASTQ-formatted files can be gzip-compressed or
                        uncompressed.
  -e EXCLUSION, --exclusion EXCLUSION
                        Path to folder containing exclusion genome(s) - those
                        you do not want signature sequences for. Genomes can
                        be in FASTA or FASTQ format. FASTA-formatted files
                        should be uncompressed, FASTQ-formatted files can be
                        gzip-compressed or uncompressed.
  -o OUTPUT_FOLDER, --output_folder OUTPUT_FOLDER
                        Path to folder where you want to store output files.
                        Folder will be created if it does not exist.
  -s KMER_SIZE, --kmer_size KMER_SIZE
                        Kmer size used to search for sequences unique to
                        inclusion. Default 31. No idea how changing this
                        affects results. TO BE INVESTIGATED.
  -t THREADS, --threads THREADS
                        Number of threads to run analysis on. Defaults to
                        number of cores on your machine.
  -pcr, --pcr           Enable to filter out inclusion kmers that have close
                        relatives in exclusion kmers.
  -k, --keep_tmpfiles   If enabled, will not clean up a bunch of (fairly)
                        useless files at the end of a run.
  -p PLASMID_FILTERING, --plasmid_filtering PLASMID_FILTERING
                        To ensure unique sequences are not plasmid-borne, a
                        FASTA-formatted database can be provided with this
                        argument. Any unique kmers that are in the plasmid
                        database will be filtered out.
  -l, --low_memory      Activate this flag to cause plasmid filtering to use
                        substantially less RAM (and go faster), at the cost of
                        some sensitivity.
  -a AMPLICON_SIZE [AMPLICON_SIZE ...], --amplicon_size AMPLICON_SIZE [AMPLICON_SIZE ...]
                        Desired size for PCR amplicons. Default 200. If you
                        want to find more than one amplicon size, enter
                        multiple, separated by spaces.

</code></pre>

<p>Additional info:</p>
<ul>
<li><code>-i, --inclusion</code>: Not too much to say about this - it's the folder where you'll want to place any genomes that you want signature sequences for. If you place more than one genome here, 
SigSeekr will look for kmers common to all input genomes and develop a signature sequence based on those. These genomes can be FASTA-formatted assemblies (recommended) in which case they must be 
uncompressed, or raw FASTQ reads, in which case they can be either uncompressed or gzip-compressed.</li>
<li><code>-e, --exclusion</code>: The collection of genomes you <em>do not</em> want your signature sequences to match to. Same file format rules as the inclusion folder. </li>
<li><code>-o, --output_folder</code>: The folder where output files will be stored. Created if it doesn't exist. Recommended that you create a new folder for each run, as outputs will be overwritten from previous runs.</li>
<li><code>-t, --threads</code>: Number of threads to run SigSeekr with. Recommended to leave at the default setting of all cores on your machine, as most programs in the SigSeekr pipeline scale very well with additional threads.</li>
<li><code>-k, --keep_tmpfiles</code>: By default, a number of fairly boring (but sometimes quite large) files are deleted at the end of a run to save on disk space. Specifying this option will keep them around if you want to inspect them more closely. Files that will be kept around with this option specified include the KMC inclusion and exclusion, and unique to inclusion databases (<code>inclusion_db</code>, <code>exclusion_db</code>, and <code>unique_to_inclusion_db</code>), FASTA files of all inclusion kmers (<code>inclusion_kmers.fasta</code>), and a bedfile showing coverage of inclusion kmers across one of the inclusion genomes specified (<code>regions_to_mask.bed</code>).</li>
<li><code>-p, --plasmid_filtering</code>: If you're looking for sequences unique to a genome, you probably don't want them on mobile elements that might not be there the next time you look. To help alleviate this potential problem, you can specify the path to a FASTA-formatted database with this option. Any inclusion kmers found in the database will be excluded from further analysis. A relatively extensive plasmid database (~9000 RefSeq plasmids spanning all of Bacteria), can be downloaded with the following command: <code>https://ndownloader.figshare.com/files/9827323 &amp;&amp; tar xf 9827323</code>.
This will create a folder called <code>databases</code> in your present working directory. Within the folder, <code>plasmid_db.fasta</code> is the plasmid database.</li>
<li><code>-l, --low_memory</code>: Using the above-mentioned plasmid database can be memory-intensive. To help alleviate that, add this flag, which will use less memory and go faster, at the cost of some sensitivity.</li>
<li><code>-pcr, --pcr</code>: Will attempt to find amplicons in your inclusion genomes that would be acceptable PCR products. By default,
searches for 200 base pair amplicons, but can be changed to any size with the <code>-a</code> option. Output file for this is <code>confirmed_amplicons_X.fasta</code>,
where X is your specified amplicon size</li>
<li><code>-a, --amplicon_size</code>: Changes the size of the PCR amplicons that SigSeekr attempts to find. If given multiple arguments separated by spaces (i.e. <code>-a 200 300 400</code>),
SigSeekr will attempt to find amplicons of all sizes specified</li>
</ul></div>
        </div>

        <footer class="col-md-12">
            <hr>
            
            <center>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</center>
        </footer>

        <script src="../js/jquery-1.10.2.min.js"></script>
        <script src="../js/bootstrap-3.0.3.min.js"></script>
        <script src="../js/highlight.pack.js"></script>
        <script src="../js/base.js"></script>
    </body>
</html>